Agarrarse a objetos

mmur07 opened the pull request:
**Cambios realizados:**
Nuevo componente:
 -AttachesToObjects: Se añade al jugador. Permite que este se agarre a otros objetos.

Cambios en clases:
 -Collider: Cambia su constructora añadiendo un booleano al final que indica si el jugador se puede agarrar a ese objeto. Si es true, crea una nueva fixture de tipo TRIGGER.

-CollisionHandler: Nuevos métodos que verifican y aplican las colisiones entre el jugador y objetos agarrables. 

**Ciclo del joint**
La lógica funciona de este modo: Empieza el step -> Detecta colisión entre jugador y objeto agarrable -> Si el jugador está pulsando la A (mando) y no está agarrado a otro objeto, la valida -> Añade los datos necesarios para crear el joint a un vector (no se puede crear un joint en mitad del step) -> Acaba el step -> Inmediatamente después, resuelve los joints (Método SolveInteractions en CollisionHandler) -> Siguiente step -> ...

**Problemas y posibles cambios**
Pensaba añadir una fixture un poco más grande que el collider original que actuara como hitbox para que el jugador pudiera agarrarse con un poco más de márgen. Esto se hace con sensores para que el jugador rebote con la hitbox pero el problema es que un sensor no puede devolver un punto de colisión al chocar con otra fixture. Esto hace que explote el programa (mover paredes estáticas...) y requeriría una forma bastante distinta de managear las colisiones (recorrer todo el vector de objetos a los que se puede agarrar el jugador en cada step para ver si colisiona con alguno). Como esta historia lleva demasiado tiempo activa y esta implementación puede funcionar, he decidido no ponerle el sensor y dejarlo con el mismo tamaño que el fixture original

mmur07 committed:
test

pa cambiar el nombre

mmur07 committed:
Primera iteración de Agarrarse a Objetos para poder empezar lo del impulso de plataformas

Cosas nuevas:

>He creado 1 nueva clase (componente) AttachableObject que se le mete al jugador en cuestión.

------------------------------------------------------
Cosas modificadas:

>He tocado la escena de prueba para que podáis ver el joint en acción.

>He añadido un getter en Collider.h para obtener un b2world y el body del jugador que necesito en AttachableObject para crear el joint. Es la manera menos fea que se me ha ocurrido de hacerlo ya que no quería pasarselo como parámetro en la constuctora. Que Álvar me castigue si lo ve necesario 😞

------------------------------------------------------
Cosas que quedan por hacer:

> Hacer que los objetos a los que te puedas agarrar tengan otro collider sensor un poco más grande que el principal al que te agarras de verdad. Pensadlo como una hitbox.

> Implementar todo esto con el inputHandler.

mmur07 committed:
Merge branch 'master' into Agarrarse_A_Objetos

mmur07 committed:
Arreglados errores de compilación al hacer el merge desde master

title

mmur07 committed:
Aplicado el InputHandler en el comportamiento de la clase

Ahora el jugador puede dejar de agarrarse a un objeto pulsando el botón A en el mando de X-Box

mmur07 committed:
Merge branch 'master' into Agarrarse_A_Objetos

mmur07 committed:
Primeros pasitos para detectar colisiones

title

mmur07 committed:
Está a punto pero ungh que dolor

Detecta las colisiones a nivel local del objeto pero una vez intenta crear el joint crashea ://

Al parecer el puntero del propio objeto da problemas??? Lo revisaré mañana.

mmur07 committed:
Sigue petando (leer descripción)

Cuando haces el enganche a pelo (PlayState.cpp línea 39) funciona bien.

Cuando haces el enganche a partir de los datos que te aporta la colisión (CollisionHandler) intenta hacer el joint con parámetros prometedores pero aborta igualmente.

mmur07 committed:
*kinda works* Pero necesito comprobarlo con físicas de verdad

mmur07 committed:
Merge branch 'master' into Agarrarse_A_Objetos

mmur07 committed:
finally fucking works

title

mmur07 committed:
ola me quiero morir salu2

ahora peta todo sijaja

mmur07 committed:
odio la vida.cpp

title

mmur07 committed:
fixes para el merge

mmur07 committed:
Merge branch 'master' into Agarrarse_A_Objetos

mmur07 committed:
Terminado

mmur07 committed:
Ya está el input de teclado

title

AdriAlvarez00 reviewed (changes requested):
Hay conflictos de merge en CollisionHandler. Parece que recibes daño aunque te agarres al chocar con un objeto, mira a ver cómo se integra con el DamageOnCollision

mmur07 committed:
Merge branch 'master' into Agarrarse_A_Objetos

mmur07 committed:
Request terminado, cambio de diseño para agarrarse a objetos y limpieza de código en PlayState

La raiz de este problema era que cuando creaba un collider al que te podías agarrar, le añadía al cuerpo una fixture adicional en la capa Trigger. Esto causaba problemas ya que, como eran del mismo tamaño, registraba 2 colisiones seguidas. La primera, la de la capa Trigger, agarraba el jugador al objeto, y la segunda, de tipo Normal, le hacía daño.

Para solucionar este problema, he creado una capa nueva: NormalAttachableObject
Cuando creas un collider al que te puedes agarrar, sobrescribe su capa original por esta nueva.
Si este diseño da problemas o es necesario expandirlo, he puesto un comentario en la constructora de collider donde explico cómo se podría solucionar.

JuanmaSalva reviewed (approved):
Ole ole lo caracole

AndresRuizB commented on src/SourceFiles/CollisionHandler.cpp:
```diff
@@ -20,30 +20,40 @@ void CollisionHandler::BeginContact(b2Contact* contact)
+	if ((contact->GetFixtureA()->GetFilterData().categoryBits == Collider::CollisionLayer::Weapon ||
+		contact->GetFixtureB()->GetFilterData().categoryBits == Collider::CollisionLayer::Weapon) &&
+		PlayerCanPickWeapon(contact, pickableObj, playerHands)) {
+		cout << "jaja si" << endl;
```
jaja no 
Poned couts significativos que si no esto es una locura

mmur07 replied:
@Pj6595 👀

mmur07 replied:
Ya debería estar en el último commit

mmur07 committed:
Arreglados couts

Pj6595 reviewed (approved):
jaja si

AndresRuizB committed:
Merge branch 'master' into Agarrarse_A_Objetos
